README.md
arduino-firmata
===============
Arduino Firmata protocol (http://firmata.org) implementation on Node.js.

- Firmata is a protocol to controll Arduino from software on PC.
- You can embed Arduino code into Node.js application.
- Support sharing an Arduino between multiple processes.
- https://github.com/shokai/node-arduino-firmata
- https://npmjs.org/package/arduino-firmata


Install
-------

    % npm install arduino-firmata


Requirements
------------

* Arduino (http://arduino.cc)
  * testing with
    * Arduino Diecimila
    * Arduino Duemillanove
    * Arduino UNO
    * Arduino Leonardo
    * Arduino Micro
    * Seeduino v2
* Arduino Standard Firmata v2.2
  * Arduino IDE -> [File] -> [Examples] -> [Firmata] -> [StandardFirmata]


Usage
-----

### Samples

- https://github.com/shokai/node-arduino-firmata/tree/master/samples

### Setup

Connect
```javascript
var ArduinoFirmata = require('arduino-firmata');
var arduino = new ArduinoFirmata();

arduino.connect(); // use default arduino
arduino.connect('/dev/tty.usb-device-name');

arduino.on('connect', function(){

  console.log("board version"+arduino.boardVersion);
  // your-code-here

});
```

Reset
```javascript
arduino.reset(callback);
```

Close
```javascript
arduino.close(callback);
```


### I/O

Digital Write
```javascript
arduino.digitalWrite(13, true, callback);
arduino.digitalWrite(13, false, callback);
```

Digital Read
```javascript
arduino.pinMode(7, ArduinoFirmata.INPUT);
console.log( arduino.digitalRead(7) ); // => true/false
```

Digital Read (event)
```javascript
arduino.pinMode(7, ArduinoFirmata.INPUT);

arduino.on('digitalChange', function(e){
  console.log("pin" + e.pin + " : " + e.old_value + " -> " + e.value);
});
```

Analog Write (PWM)
```
setInterval(function(){
  var an = Math.random()*255; // 0 ~ 255
  arduino.analogWrite(9, an, callback);
}, 100);
```

Analog Read
```javascript
console.log( arduino.analogRead(0) ); // => 0 ~ 1023
```

Analog Read (event)
```javascript
arduino.on('analogChange', function(e){
  console.log("pin" + e.pin + " : " + e.old_value + " -> " + e.value);
});
```

Servo Motor
```javascript
setInterval(function(){
  var angle = Math.random()*180; // 0 ~ 180
  arduino.servoWrite(11, angle, callback);
}, 1000);
```

### Sysex

- http://firmata.org/wiki/V2.1ProtocolDetails#Sysex_Message_Format
- https://github.com/shokai/node-arduino-firmata/tree/master/samples/sysex

Send
```javascript
arduino.sysex(0x01, [13, 5, 2], callback);  // command, data_array, callback
```

Register Sysex Event
```javascript
arduino.on('sysex', function(e){
  console.log("command : " + e.command);
  console.log(e.data);
});
```


Develop
-------

    % npm install
    % npm run build
    # or
    % npm run watch

Test
----

### Install SysexLedBlinkFirmata into Arduino

* https://github.com/shokai/node-arduino-firmata/blob/master/samples/sysex/StandardFirmataWithLedBlink/StandardFirmataWithLedBlink.ino


### Run Test

    % npm test


Contributing
------------
1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


History.txt
=== 0.3.4 2016-02-18

* add serialport for RaspberryPi
  * thank you for contributing @arrowmeiwaracing

=== 0.3.3 2015-08-10

* not commit compiled js in Git repos
* not publish coffee script to npmjs.com


=== 0.3.2 2014-10-02

* use latest npm libs
  * serialport, debug, eventemitter2
  * thank you for suggestion @bigmonkeyboy

* coffee-script v1.8.x to compile js

=== 0.3.1 2014-05-08

* bugfix pin number of event "digitalChange"  #12

=== 0.3.0 2014-01-31

* add supported boards Arduino diecimila, duemillanove and seeduino  #4

=== 0.2.1 2014-01-29

* fix servoWrite sample and README
* callback after digitalWrite, analogWrite, servoWrite, reset and sysex  #11

=== 0.2.0 2014-01-28

* implement sysex(cmd, data_array)  #6

=== 0.1.0 2014-01-27

* implement on("analogChange", callback)  #9
* implement on("digitalChange", callback)  #8
* fix socket.io sample

=== 0.0.5 2014-01-27

* performance improvement (write bytes at once)
* add socket.io sample  #7

=== 0.0.4 2014-01-27

* add tests  #3
* bugfix digitalRead()

=== 0.0.3 2014-01-27

* fix coffee-script  #2
  * thank you for contributing @geta6

=== 0.0.2 2014-01-26

* use SerialPort.list() instead of fs.readdir()  #1
  * thank you for suggestion @rwaldron

=== 0.0.1 2014-01-26

* first release
* implement digitalWrite, digitalRead, analogWrite, analogRead, servoWrite


lib/arduino-firmata.js
(function() {
  'use strict';
  var ArduinoFirmata, SerialPort, debug, events, exports, serialport,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  events = require('eventemitter2');

  SerialPort = (serialport = require('serialport')).SerialPort;

  debug = require('debug')('arduino-firmata');

  exports = module.exports = ArduinoFirmata = (function(_super) {
    __extends(ArduinoFirmata, _super);

    ArduinoFirmata.Status = {
      CLOSE: 0,
      OPEN: 1
    };

    ArduinoFirmata.INPUT = 0;

    ArduinoFirmata.OUTPUT = 1;

    ArduinoFirmata.ANALOG = 2;

    ArduinoFirmata.PWM = 3;

    ArduinoFirmata.SERVO = 4;

    ArduinoFirmata.SHIFT = 5;

    ArduinoFirmata.I2C = 6;

    ArduinoFirmata.LOW = 0;

    ArduinoFirmata.HIGH = 1;

    ArduinoFirmata.MAX_DATA_BYTES = 32;

    ArduinoFirmata.DIGITAL_MESSAGE = 0x90;

    ArduinoFirmata.ANALOG_MESSAGE = 0xE0;

    ArduinoFirmata.REPORT_ANALOG = 0xC0;

    ArduinoFirmata.REPORT_DIGITAL = 0xD0;

    ArduinoFirmata.SET_PIN_MODE = 0xF4;

    ArduinoFirmata.REPORT_VERSION = 0xF9;

    ArduinoFirmata.SYSTEM_RESET = 0xFF;

    ArduinoFirmata.START_SYSEX = 0xF0;

    ArduinoFirmata.END_SYSEX = 0xF7;

    ArduinoFirmata.list = function(callback) {
      return serialport.list(function(err, ports) {
        var devices, port, _i, _len;
        if (err) {
          return callback(err);
        }
        devices = [];
        for (_i = 0, _len = ports.length; _i < _len; _i++) {
          port = ports[_i];
          if (/usb|acm|com\d+/i.test(port.comName)) {
            devices.push(port.comName);
          }
        }
        return callback(null, devices);
      });
    };

    function ArduinoFirmata() {
      this.status = ArduinoFirmata.Status.CLOSE;
      this.wait_for_data = 0;
      this.execute_multi_byte_command = 0;
      this.multi_byte_channel = 0;
      this.stored_input_data = [];
      this.parsing_sysex = false;
      this.sysex_bytes_read = 0;
      this.digital_output_data = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
      this.digital_input_data = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
      this.analog_input_data = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
      this.boardVersion = null;
    }

    ArduinoFirmata.prototype.isOldArduinoDevice = function() {
      return /usbserial|USB/.test(this.serialport_name);
    };

    ArduinoFirmata.prototype.connect = function(serialport_name, opts) {
      this.serialport_name = serialport_name;
      if (opts == null) {
        opts = {
          baudrate: 57600
        };
      }
      opts.parser = serialport.parsers.raw;
      if (!this.serialport_name) {
        ArduinoFirmata.list((function(_this) {
          return function(err, devices) {
            return _this.connect(devices[0], opts);
          };
        })(this));
        return this;
      }
      this.once('boardReady', function() {
        var io_init_wait;
        debug('boardReady');
        io_init_wait = this.isOldArduinoDevice() ? (debug("old arduino device found " + this.serialport_name), 3000) : (debug("new arduino device found " + this.serialport_name), 100);
        debug("wait " + io_init_wait + "(msec)");
        return setTimeout((function(_this) {
          return function() {
            var i, _i, _j;
            for (i = _i = 0; _i < 6; i = ++_i) {
              _this.write([ArduinoFirmata.REPORT_ANALOG | i, 1]);
            }
            for (i = _j = 0; _j < 2; i = ++_j) {
              _this.write([ArduinoFirmata.REPORT_DIGITAL | i, 1]);
            }
            debug('init IO ports');
            return _this.emit('connect');
          };
        })(this), io_init_wait);
      });
      this.serialport = new SerialPort(this.serialport_name, opts);
      this.serialport.once('open', (function(_this) {
        return function() {
          var cid;
          cid = setInterval(function() {
            debug('request REPORT_VERSION');
            return _this.write([ArduinoFirmata.REPORT_VERSION]);
          }, 500);
          _this.once('boardVersion', function(version) {
            clearInterval(cid);
            _this.status = ArduinoFirmata.Status.OPEN;
            return _this.emit('boardReady');
          });
          return _this.serialport.on('data', function(data) {
            var byte, _i, _len, _results;
            _results = [];
            for (_i = 0, _len = data.length; _i < _len; _i++) {
              byte = data[_i];
              _results.push(_this.process_input(byte));
            }
            return _results;
          });
        };
      })(this));
      return this;
    };

    ArduinoFirmata.prototype.isOpen = function() {
      return this.status === ArduinoFirmata.Status.OPEN;
    };

    ArduinoFirmata.prototype.close = function(callback) {
      this.status = ArduinoFirmata.Status.CLOSE;
      return this.serialport.close(callback);
    };

    ArduinoFirmata.prototype.reset = function(callback) {
      return this.write([ArduinoFirmata.SYSTEM_RESET], callback);
    };

    ArduinoFirmata.prototype.write = function(bytes, callback) {
      return this.serialport.write(bytes, callback);
    };

    ArduinoFirmata.prototype.sysex = function(command, data, callback) {
      var write_data;
      if (data == null) {
        data = [];
      }
      data = data.map(function(i) {
        return i & 0x7f;
      });
      write_data = [ArduinoFirmata.START_SYSEX, command].concat(data, [ArduinoFirmata.END_SYSEX]);
      return this.write(write_data, callback);
    };

    ArduinoFirmata.prototype.pinMode = function(pin, mode, callback) {
      switch (mode) {
        case true:
          mode = ArduinoFirmata.OUTPUT;
          break;
        case false:
          mode = ArduinoFirmata.INPUT;
      }
      return this.write([ArduinoFirmata.SET_PIN_MODE, pin, mode], callback);
    };

    ArduinoFirmata.prototype.digitalWrite = function(pin, value, callback) {
      var port_num;
      this.pinMode(pin, ArduinoFirmata.OUTPUT);
      port_num = (pin >>> 3) & 0x0F;
      if (value === 0 || value === false) {
        this.digital_output_data[port_num] &= ~(1 << (pin & 0x07));
      } else {
        this.digital_output_data[port_num] |= 1 << (pin & 0x07);
      }
      return this.write([ArduinoFirmata.DIGITAL_MESSAGE | port_num, this.digital_output_data[port_num] & 0x7F, this.digital_output_data[port_num] >>> 7], callback);
    };

    ArduinoFirmata.prototype.analogWrite = function(pin, value, callback) {
      value = Math.floor(value);
      this.pinMode(pin, ArduinoFirmata.PWM);
      return this.write([ArduinoFirmata.ANALOG_MESSAGE | (pin & 0x0F), value & 0x7F, value >>> 7], callback);
    };

    ArduinoFirmata.prototype.servoWrite = function(pin, angle, callback) {
      this.pinMode(pin, ArduinoFirmata.SERVO);
      return this.write([ArduinoFirmata.ANALOG_MESSAGE | (pin & 0x0F), angle & 0x7F, angle >>> 7], callback);
    };

    ArduinoFirmata.prototype.digitalRead = function(pin) {
      return ((this.digital_input_data[pin >>> 3] >>> (pin & 0x07)) & 0x01) > 0;
    };

    ArduinoFirmata.prototype.analogRead = function(pin) {
      return this.analog_input_data[pin];
    };

    ArduinoFirmata.prototype.process_input = function(input_data) {
      var analog_value, command, diff, i, old_analog_value, stat, sysex_command, sysex_data, _i, _results;
      if (this.parsing_sysex) {
        if (input_data === ArduinoFirmata.END_SYSEX) {
          this.parsing_sysex = false;
          sysex_command = this.stored_input_data[0];
          sysex_data = this.stored_input_data.slice(1, this.sysex_bytes_read);
          return this.emit('sysex', {
            command: sysex_command,
            data: sysex_data
          });
        } else {
          this.stored_input_data[this.sysex_bytes_read] = input_data;
          return this.sysex_bytes_read += 1;
        }
      } else if (this.wait_for_data > 0 && input_data < 128) {
        this.wait_for_data -= 1;
        this.stored_input_data[this.wait_for_data] = input_data;
        if (this.execute_multi_byte_command !== 0 && this.wait_for_data === 0) {
          switch (this.execute_multi_byte_command) {
            case ArduinoFirmata.DIGITAL_MESSAGE:
              input_data = (this.stored_input_data[0] << 7) + this.stored_input_data[1];
              diff = this.digital_input_data[this.multi_byte_channel] ^ input_data;
              this.digital_input_data[this.multi_byte_channel] = input_data;
              if (this.listeners('digitalChange').length > 0) {
                _results = [];
                for (i = _i = 0; _i <= 13; i = ++_i) {
                  if (((0x01 << i) & diff) > 0) {
                    stat = (input_data & diff) > 0;
                    _results.push(this.emit('digitalChange', {
                      pin: i + this.multi_byte_channel * 8,
                      value: stat,
                      old_value: !stat
                    }));
                  } else {
                    _results.push(void 0);
                  }
                }
                return _results;
              }
              break;
            case ArduinoFirmata.ANALOG_MESSAGE:
              analog_value = (this.stored_input_data[0] << 7) + this.stored_input_data[1];
              old_analog_value = this.analogRead(this.multi_byte_channel);
              this.analog_input_data[this.multi_byte_channel] = analog_value;
              if (old_analog_value !== analog_value) {
                return this.emit('analogChange', {
                  pin: this.multi_byte_channel,
                  value: analog_value,
                  old_value: old_analog_value
                });
              }
              break;
            case ArduinoFirmata.REPORT_VERSION:
              this.boardVersion = "" + this.stored_input_data[1] + "." + this.stored_input_data[0];
              return this.emit('boardVersion', this.boardVersion);
          }
        }
      } else {
        if (input_data < 0xF0) {
          command = input_data & 0xF0;
          this.multi_byte_channel = input_data & 0x0F;
        } else {
          command = input_data;
        }
        if (command === ArduinoFirmata.START_SYSEX) {
          this.parsing_sysex = true;
          return this.sysex_bytes_read = 0;
        } else if (command === ArduinoFirmata.DIGITAL_MESSAGE || command === ArduinoFirmata.ANALOG_MESSAGE || command === ArduinoFirmata.REPORT_VERSION) {
          this.wait_for_data = 2;
          return this.execute_multi_byte_command = command;
        }
      }
    };

    return ArduinoFirmata;

  })(events.EventEmitter2);

}).call(this);
